libraries/chain/include/eosio/chain/log_catalog.hpp

```diff
@@ -0,0 +1,248 @@
+      if (collection.empty() || block_num < collection.begin()->first)
+         return "";
+      auto it = --collection.upper_bound(block_num);
+
+      if (block_num <= it->second.last_block_num)
+         return it->second.filename_base;
+      return "";
+   }
+
+   bool set_active_item(uint32_t block_num, mapmode mode = mapmode::readonly) {
+      try {
+         if (active_index != npos) {
+            auto active_item = collection.nth(active_index);
+            if (active_item->first <= block_num && block_num <= active_item->second.last_block_num &&
+                log_data.flags() == mode) {
+               return true;
+            }
+         }
+         if (collection.empty() || block_num < collection.begin()->first)
+            return false;
+
+         auto it = --collection.upper_bound(block_num);
+
+         if (block_num <= it->second.last_block_num) {
+            auto name = it->second.filename_base;
+            log_data.open(name.replace_extension("log"), mode);
+            log_index.open(name.replace_extension("index"));
+            this->active_index = collection.index_of(it);
+            return true;
+         }
+         return false;
+      } catch (...) {
+         this->active_index = npos;
+         return false;
+      }
+   }
+
+   std::pair<fc::datastream<const char*>, uint32_t> ro_stream_for_block(uint32_t block_num) {
+      if (set_active_item(block_num, mapmode::readonly)) {
+         auto pos = log_index.nth_block_position(block_num - log_data.first_block_num());
+         return std::make_pair(log_data.ro_stream_at(pos), log_data.version());
+      }
+      return {fc::datastream<const char*>(nullptr, 0), static_cast<uint32_t>(0)};
+   }
+
+   std::pair<fc::datastream<char*>, uint32_t> rw_stream_for_block(uint32_t block_num) {
+      if (set_active_item(block_num, mapmode::readwrite)) {
+         auto pos = log_index.nth_block_position(block_num - log_data.first_block_num());
+         return std::make_pair(log_data.rw_stream_at(pos), log_data.version());
+      }
+      return {fc::datastream<char*>(nullptr, 0), static_cast<uint32_t>(0)};
+   }
+
+   std::optional<block_id_type> id_for_block(uint32_t block_num) {
+      if (set_active_item(block_num, mapmode::readonly)) {
+         auto pos = log_index.nth_block_position(block_num - log_data.first_block_num());
+         return log_data.block_id_at(pos);
+      }
+      return {};
+   }
+
+   static void rename_if_not_exists(bfs::path old_name, bfs::path new_name) {
+      if (!bfs::exists(new_name)) {
+         bfs::rename(old_name, new_name);
+      }
+      else {
+         bfs::remove(old_name);
+         wlog("${new_name} already exists, just remove ${old_name}", ("old_name", old_name.string())("new_name", new_name.string()));
+      }
+   }
+
+   static void rename_bundle(bfs::path orig_path, bfs::path new_path) {
+      rename_if_not_exists(orig_path.replace_extension(".log"), new_path.replace_extension(".log"));
```

Comment:
It shouldn't matter. The design was to cover the case when the archive or retaining dir is a network shared directory like (NFS mount), where multiple nodeos can write to it. If the file already exists because another nodeos instance has written to it. We should be able to ignore it.